### DISCUSSION TITLE
Janusgraph read query fails with "cannot write to closing transport"

### DISCUSSION THREAD

**Discussion by sharma-kunal** [JanusGraph/janusgraph#4035] [Q&A] [answered]
I am using `Janusgraph: 0.6.3` and `gremlinpython=3.5.5`

I have a Flask API and I create a Janusgraph Connection Instance when my API is started like this,

```
def connect(self):
    try:
        connection = DriverRemoteConnection(url=f"ws://{self.properties.get('ENDPOINT')}/gremlin",
                                            traversal_source=self.properties.get('TRAVERSAL_SOURCE'))
        self.g = traversal().withRemote(connection)
    except Exception as e:
        LoggerUtil.logger.error(f"Connection with Janusgraph Refused with msg: {e}")
```

Now, I run my read queries in this graph like,

```
g.V().hasLabel("<label>").has("name", "<name>").range_(0,10).toList()
```

and more.

Now, these queries work fine but sometimes they fail with the error

```
Connection was already closed.
```

and sometimes with
```
cannot write to closing transport
```

What does this error specify ? Do I need to close the transport or transaction explicitly after I run any read query on Graph ?

**Comment by FlorianHockmann** [comment]
How is this discussion different to what you wrote in #4022? Please don't create duplicate issues. It just makes it even harder for people who could possibly help you to keep up with new questions asked by users.

**Reply by sharma-kunal** [reply]
Sorry for the duplicate issue. I'll close #4022.

**Comment by hadoopmarc** [comment]
Is the connect() function called for every request to the flask app? This would not be OK. You would have to establish the connection at app startup and only reconnect if the connection would have timed out (this a long time, it is a websocket connection).

In addition, to Florian's remark: it is possible to edit the original question if you want to add details.

**Reply by sharma-kunal** [reply]
No, the `connect()` function is only called once when the Flask App Starts and never again. So, as per the error message `Connection was already closed`, I am unable to understand which connection it is referring to. and after the error message, if I send the request after some time (1 or 2 sec later), it again works fine.

**Reply by hadoopmarc** [reply]
As far as I can see, Gremlin Server should keep the websocket connection to flask open, unless you would have changed default yaml properties for idleConnectionTimeout and keepAliveInterval. Are there any other mechanisms that could mess with the connection? Firewalls, load managers, gunicorn, etc. Maybe also try from a python shell if that can maintain the connection. What timescales are you talking: after how long/ how often the connection is lost?

**Comment by sharma-kunal** [comment]
No, I have not changed any default properties of Janusgraph. I have checked firewalls etc, and they seem to be causing no issue with Janusgraph connectivity. I am guessing this might be some gremlinpython version issue. I am currently using gremlinpython 3.5.5 with Janusgraph 0.6.3. I have updated gremlinpython to 3.6.0 and testing now. The problem here is the issue is not persistent. Sometimes it comes very often, but sometimes it never comes. So, it's hard to debug. What do you suggest? Is there any specific version of gremlinpython for Janusgraph 0.6.3.

**Reply by hadoopmarc** [reply]
gremlinpython-3.5.5 goes with janusgraph-0.6.3, see: https://github.com/JanusGraph/janusgraph/releases

**Reply by hadoopmarc** [reply]
This might help: https://stackoverflow.com/questions/63234523/gremlin-python-server-disconnected-please-try-to-reconnect-error

**Comment by sharma-kunal** [comment, marked as answer]
I changed my gremlinpython version from 3.5.5 to 3.6.0 and it is working fine now, I am not sure what the issue was, but updating my gremlinpython to 3.6.0 helped. Thanks for the help @hadoopmarc 

@hadoopmarc The release page you shared above, even on that page for Janusgraph 0.6.3 version, gremlinpython they are using is 3.5.7, here: https://docs.janusgraph.org/interactions/connecting/python/
